src/SSTIO.cpp

```diff
@@ -0,0 +1,514 @@
+                if (err == -1) {
+                    throw Exception("SSTIOImp::read_batch() mbox read failed",
+                                    errno, __FILE__, __LINE__);
+                }
+            }
+        }
+        if (!m_mmio_read_interfaces.empty()) {
+            m_mmio_read_batch = ioctl_structs_from_vector<sst_mmio_interface_batch_s>(
+                m_mmio_read_interfaces);
+
+            for (auto &batch : m_mmio_read_batch) {
+                int err = ioctl(m_fd, GEOPM_IOC_SST_MMIO, batch.get());
+                if (err == -1) {
+                    throw Exception("SSTIOImp::read_batch() mmio read failed",
+                                    errno, __FILE__, __LINE__);
+                }
+            }
+        }
+    }
+
+    uint64_t SSTIOImp::sample(int batch_idx) const
+    {
+        const auto& interface = m_added_interfaces[batch_idx];
+        uint64_t sample_value;
+        if (interface.first == MMIO) {
+            size_t mmio_batch_idx = interface.second / m_batch_command_limit;
+            size_t mmio_interface_idx = interface.second % m_batch_command_limit;
+            sample_value = m_mmio_read_batch[mmio_batch_idx]->interfaces[mmio_interface_idx].value;
+        }
+        else {
+            size_t mbox_batch_idx = interface.second / m_batch_command_limit;
+            size_t mbox_interface_idx = interface.second % m_batch_command_limit;
+            sample_value = m_mbox_read_batch[mbox_batch_idx]->interfaces[mbox_interface_idx].read_value;
+        }
+        return sample_value;
+    }
+
+    void SSTIOImp::write_batch(void)
+    {
+        if (!m_mbox_write_interfaces.empty()) {
+            m_mbox_write_batch = ioctl_structs_from_vector<sst_mbox_interface_batch_s>(
+                m_mbox_rmw_interfaces);
+
+            for (auto &batch : m_mbox_write_batch) {
+                // Read existing value (TODO: only need if not whole mask write)
+                int err = ioctl(m_fd, GEOPM_IOC_SST_MBOX, batch.get());
+                if (err == -1) {
+                    throw Exception("sstioimp::write_batch() pre-write mbox read failed",
+                                    errno, __FILE__, __LINE__);
+                }
+
+                // Modify the existing values with the adjusted values, using
+                // the buffer that contains the mailbox write locations (which
+                // may be different from the read locations for some controls)
+                for (size_t i = 0; i < batch->num_entries; ++i) {
+                    // Mask the read so we only propagate the bits that we are
+                    // supposed to read. Mask the write so we only update the
+                    // adjusted bits.
+                    m_mbox_write_interfaces[i].write_value |=
+                        ~m_mbox_rmw_write_masks[i] &
+                        (batch->interfaces[i].read_value &
+                         m_mbox_rmw_read_masks[i]);
```

Comment:
is there a check somewhere that these two masks don't overlap?  what are the bits not covered by either mask?